cli/src/complete.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020-2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap::FromArgMatches as _;
+use clap_complete::CompletionCandidate;
+use jj_lib::workspace::DefaultWorkspaceLoaderFactory;
+use jj_lib::workspace::WorkspaceLoaderFactory as _;
+
+use crate::cli_util::find_workspace_dir;
+use crate::cli_util::resolve_aliases;
+use crate::cli_util::resolve_default_command;
+use crate::cli_util::GlobalArgs;
+use crate::config::default_config;
+use crate::config::LayeredConfigs;
+use crate::ui::Ui;
+
+pub fn local_bookmarks() -> Vec<CompletionCandidate> {
+    jj().arg("bookmark")
+        .arg("list")
+        .arg("--template")
+        .arg(r#"if(!remote, name ++ "\n")"#)
+        .output()
+        .map(|output| output.stdout)
+        .ok()
+        .and_then(|stdout| String::from_utf8(stdout).ok())
+        .map(|stdout| stdout.lines().map(CompletionCandidate::new).collect())
+        .unwrap_or_default()
+}
+
+/// Shell out to jj during dynamic completion generation
+///
+/// This is necessary because dynamic completion code needs to be aware of
+/// global configuration like custom storage backends. Dynamic completion
+/// code via clap_complete doesn't accept arguments, so they cannot be passed
+/// that way. Another solution would've been to use global mutable state, to
+/// give completion code access to custom backends. Shelling out was chosen as
+/// the preferred method, because it's more maintainable and the performance
+/// requirements of completions aren't very high.
+fn jj() -> std::process::Command {
+    let mut command = std::process::Command::new(
+        std::env::current_exe().unwrap_or(std::path::PathBuf::from("jj")),
```

Comment:
nit: it's probably better to print error message to stderr instead of assuming the executable name is `jj`.